### PR TITLE
Add some more FIRRTL annotation documentation

### PR DESCRIPTION
**Pull request by youngar**
This moves the annotation documentation out of the FIRRTL rational and in to its own document.  It expands the target section a little bit with some details from a Google Docs document.  At the end of the document is reference style material for each annotation.  This works pretty well for the simple annotations, but for GrandCentral each annotation had to have several tables to describe the format. Details were collected from the Scala source code and our tests.

The Grand Central Views documentation could use some more clarifying details.  One particular place I am unclear is `UnknownGroundType` in the GrandCentralView annotation; is every AugmentedGroundType an UnknownGroundType?

The Scala format specifier was previously used to get decent highlighting for FIRRTL code, but this does not render well on the CIRCT website.

### CODE REVIEW

**Comment by youngar** [comment]
Probably the easiest way to view it is here: https://github.com/youngar/circt/blob/firrtl-doc-annotations/docs/FIRRTLAnnotations.md

File: docs/FIRRTLAnnotations.md [hunk 1]
```diff
@@ -0,0 +1,717 @@
+# FIRRTL Annotations
+
+The SFC provides a mechanism to encode arbitrary metadata and associate it with
```
Comment (lattner) [inline docs/FIRRTLAnnotations.md:3]: Super nit, but I'd say "Scala FIRRTL Compiler (SFC)" to make the doc more self contained.

File: docs/FIRRTLAnnotations.md [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+This can be equivalently, in CIRCT, expressed as:
+
+```firrtl
+circuit Foo: %[[{"target":"~Foo|Foo","hello":"world"}]]
+  module Foo:
+    skip
+```
+
+During parsing, annotations are "scattered" into the MLIR representation as
+operation or argument attributes.  As an example of this, the above parses into
```
Comment (lattner) [inline docs/FIRRTLAnnotations.md:172]: I'd recommend s/argument/port/ or BlockArgument for specificity

File: docs/FIRRTLAnnotations.md [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+the attribute.  The `LowerTypes` pass then uses this reference information to
+attach annotation metadata to only the _lowered_ portion of a targeted circuit
+component.
+
+Annotations are expected to be fully removed via custom transforms, conversion
+to other MLIR operations, or dropped.  E.g., the `ModuleInliner` pass removes
+`firrtl.passes.InlineAnnotation` by inlining annotated modules or instances.
+JSON Annotations map to the builtin MLIR attributes. An annotation is
+implemented using a DictionaryAttr, which holds the class, target, any
+annotation specific data.
```
Comment (lattner) [inline docs/FIRRTLAnnotations.md:195]: Mention that unhandled annotations are warned about in Lower2HW?

**Review by lattner — APPROVED** [review]
This is amazingly great, thank you for writing this up!

**Review by seldridge — APPROVED** [review]
This looks awesome, @youngar.

As the FIRRTL Dialect-specific documentation expands, do you think we should start a `FIRRTL` or `FIRRTL Dialect` directory?

File: docs/FIRRTLAnnotations.md (hunk 1 above)
Comment (seldridge) [reply docs/FIRRTLAnnotations.md:3]: Agree. A reasonable policy is to define everything the first time in every separate doc.

File: docs/FIRRTLAnnotations.md [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+
+Example:
+```json
+{
+  "class": "firrtl.transforms.BlackBoxInlineAnno",
+  "target": "~Foo|Foo",
+  "name": "blackbox-inline.v",
+  "text": "module ExtInline(); endmodule\n"
+}
+```
```
Comment (seldridge) [inline docs/FIRRTLAnnotations.md:223]: Providing more info on an offline conversation @youngar and I had about this:

There a lot of ways to represent this including tables, prose, JSON examples, JSON schema, other schemas, etc.  Any documentation is better than no documentation and I'm in favor of just keeping this as this solid looking table + prose + example. 

Longer term, there is a lot of information here that _could_ be captured in ODS and generated, including the documentation.  E.g., the unpacking logic for this annotation could be derived from an ODS representation (https://github.com/llvm/circt/blob/6e8be3dc0103c087deb6c97fa072e56b4b2e6794/lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp#L212).

File: docs/FIRRTLAnnotations.md [hunk 5]
```diff
@@ -657,7 +657,7 @@ This key allows the creation of a FIRRTL literal.
 
 The DataTapsAnnotation is a collection of all the data taps in a circuit. This
 will cause a data tap module to be emitted.  The DataTapsAnnotation implies
-DontTouchAnnotation.
+DontTouchAnnotation on the `blackbox` target.
```
Comment (seldridge) [inline docs/FIRRTLAnnotations.md:660]: This is technically on all _keys_ and not on the blackBox target.  I.e., anything that is a key should not be removed from the circuit.

Comment (youngar) [reply docs/FIRRTLAnnotations.md:660]: Yeah, looks like you're right, I was reading the code wrong.  Looks like we are attaching the `DontTouchAnnotation` on to the [blackBox](https://github.com/llvm/circt/blob/main/lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp#L673), almost every blackbox port (targeted by a key), and some ModuleTargets, but upstream it's only the `ReferenceDataTapKey.source` that gets DontTouch (EmitDataTaps.scala:115)? 

```scala
  def dontTouches: Seq[ReferenceTarget] = {
    keys.collect {
      case key: ReferenceDataTapKey => key.source
    }
```

Any idea what would make the most sense to document here?

The View annotation doesn't document its don't touches on every GroundType, which I will also add:
```scala
// ViewAnnotation
  def collectDontTouches(tpe: AugmentedType): Seq[ReferenceTarget] = tpe match {
    case b: AugmentedBundleType =>
      b.elements.flatMap { field =>
        collectDontTouches(field.tpe)
      }
    case v: AugmentedVectorType => v.elements.flatMap(collectDontTouches(_))
    case g: AugmentedGroundType => Seq(g.ref)
    case other => Seq.empty
  }
```

Comment (youngar) [reply docs/FIRRTLAnnotations.md:660]: Updated change only mentions what SFC puts don't touch on (`ReferenceDataTapKey` and `AugmentedGeroundType`). I am not sure if I should mention `DataTapModuleSignalKey.module` adding DontTouch.  Most of the other `DontTouches` we add (to the blackbox+ports) seem like an implementation detail and not worth mentioning.   WDYT?

**Review by fabianschuiki — APPROVED** [review]
This is *extremely* nice! Looking forward to seeing this as part of the [deployed documentation](https://circt.llvm.org/docs/).

File: docs/FIRRTLAnnotations.md (hunk 5 above)
Comment (seldridge) [reply docs/FIRRTLAnnotations.md:660]: Good catch. It looks we're being overzealous with dont touch relative to the SFC implementation. We may need to walk that back.

**Comment by youngar** [comment]
Merging this now, but I'm sure there is still a lot of room for improvement. @seldridge pointed me at @azidar's [dissertation](https://www2.eecs.berkeley.edu/Pubs/TechRpts/2019/EECS-2019-168.html) section 4, which has even more information about how annotations work. There is lots of interesting stuff in there!

**Pull request merged**